From 95dd2f4abcd95ecc9253844176694b92544f111b Mon Sep 17 00:00:00 2001
From: Yuxuan Shui <yshui@codeweavers.com>
Date: Tue, 1 Jul 2025 21:28:57 +0100
Subject: [PATCH] services: Fix use-after-free during shutdown.

Since rpc processes requests on a work queue, it might hold a referecen to the
connection while the main thread is shutting down and freeing the database.
When the work item completes, it will release the connection which in turn
calls services.exe's rundown function which uses the freed database.
---
 programs/services/services.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/programs/services/services.c b/programs/services/services.c
index dff3cd60040..5e583baf1ac 100644
--- a/programs/services/services.c
+++ b/programs/services/services.c
@@ -493,6 +493,12 @@ static void scmdatabase_wait_terminate(struct scmdatabase *db)
         list_add_tail(&db->processes, &process->entry);
         release_process(process);
     }
+    while (!list_empty(&db->services))
+    {
+        scmdatabase_unlock(db);
+        Sleep(500);
+        scmdatabase_lock(db);
+    }
     scmdatabase_unlock(db);
 }
 
-- 
GitLab

